dataland-email-service/src/main/kotlin/org/dataland/datalandemailservice/services/EmailSubscriptionTracker.kt

```diff
@@ -0,0 +1,52 @@
+package org.dataland.datalandemailservice.services
+
+import org.dataland.datalandemailservice.entities.EmailSubscriptionEntity
+import org.dataland.datalandemailservice.repositories.EmailSubscriptionRepository
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+import org.springframework.transaction.annotation.Transactional
+import java.util.UUID
+
+/**
+ * Service responsible for managing email subscriptions.
+ */
+@Component
+class EmailSubscriptionTracker(
+    @Autowired private val emailSubscriptionRepository: EmailSubscriptionRepository,
+) {
+    /**
+     * Inserts a new email subscription if one does not already exist and returns its UUID.
+     *
+     * This method checks if an email subscription exists for the given email address.
+     * If it does not exist, a new subscription is created with `isSubscribed` set to `true`.
+     * The method returns the UUID of the subscription if it is active (subscribed),
+     * otherwise returns `null`.
+     *
+     * @param emailAddress The email address to subscribe.
+     * @return The UUID of the active/inactive subscription,
+     * or the UUID of the newly created entity if no subscription existed.
+     */
+    @Transactional
+    fun insertSubscriptionEntityIfNeededAndReturnUuid(emailAddress: String): UUID {
+        val entity =
+            emailSubscriptionRepository.findByEmailAddress(emailAddress)
+                ?: emailSubscriptionRepository.save(
+                    EmailSubscriptionEntity(
+                        emailAddress = emailAddress,
+                        isSubscribed = true,
+                    ),
+                )
+
+        return entity.uuid
+    }
+
+    /**
+     * Checks if the specified email address is subscribed.
+     *
+     * This function queries the [EmailSubscriptionRepository] to determine whether the
+     * provided [emailAddress] is currently subscribed. It returns `true` if the email
+     * is subscribed, `false` if it is not subscribed, and `null` if the email address
+     * does not exist in the repository.
+     */
+    fun emailIsSubscribed(emailAddress: String): Boolean? = emailSubscriptionRepository.findByEmailAddress(emailAddress)?.isSubscribed
```

Comment:
Also: Why not return `false` if emailAddress couldn't be found? If it couldn't be found, it is obviously not subscribed

Reply:
As explained above, this case is actually mapped to true. The possible states for the email subscription can be either subscribed, unsubscribed or unknown, this is reflected by the Boolean? type.